### PR TITLE
Add a new task def attribute to support task poll timeout

### PR DESCRIPTION
**Pull request by nbraquart**
As explained here: https://github.com/Netflix/conductor/issues/407

### CODE REVIEW

File: core/src/main/java/com/netflix/conductor/core/execution/DeciderService.java [hunk 1]
```diff
@@ -524,23 +524,38 @@ void checkTaskTimeout(TaskDef taskDef, Task task) {
...
-        long timeout = 1000L * taskDef.getTimeoutSeconds();
-        long now = System.currentTimeMillis();
-        long elapsedTime = now - (task.getStartTime() + ((long) task.getStartDelayInSeconds() * 1000L));
+        String reason = null;
+        final long now = System.currentTimeMillis();
+        final long startDelay = 1000L * task.getStartDelayInSeconds();
 
-        if (elapsedTime < timeout) {
-            return;
+        if (task.getStatus().equals(SCHEDULED) && taskDef.getPollTimeoutSeconds() > 0) {
```
Comment (Jiehong) [inline core/src/main/java/com/netflix/conductor/core/execution/DeciderService.java]: if `taskDef.getPollTimeoutSeconds()` is `null` (the case for all tasks defined before that functionality has been merged, and it's optional), this check is a bit weird.

Perhaps only checking it's not null is enough instead.

File: common/src/main/java/com/netflix/conductor/common/metadata/tasks/TaskDef.java [hunk 2]
```diff
@@ -113,6 +113,9 @@
 	@Email(message = "ownerEmail should be valid email address")
 	private String ownerEmail;
 
+	@ProtoField(id = 19)
+	private Integer pollTimeoutSeconds;
```
Comment (Jiehong) [inline common/src/main/java/com/netflix/conductor/common/metadata/tasks/TaskDef.java]: This field shouldn't be allowed to be negative. Could you please add a `@Min` like for `responseTimeoutSeconds` ?

File: core/src/main/java/com/netflix/conductor/core/execution/DeciderService.java (hunk 1 above)
Comment (andrea11) [reply core/src/main/java/com/netflix/conductor/core/execution/DeciderService.java]: I think both need to be checked. We can assume (to write down on the doc) that if pollTimeoutSeconds is set to 0, no checkTaskTimeout should be performed in this case

File: core/src/main/java/com/netflix/conductor/core/execution/DeciderService.java [hunk 3]
```diff
@@ -524,23 +524,38 @@ void checkTaskTimeout(TaskDef taskDef, Task task) {
...
         }
-        if (task.getStatus().isTerminal() || taskDef.getTimeoutSeconds() <= 0 || task.getStartTime() <= 0) {
+        if (task.getStatus().isTerminal()) {
             return;
         }
 
-        long timeout = 1000L * taskDef.getTimeoutSeconds();
-        long now = System.currentTimeMillis();
-        long elapsedTime = now - (task.getStartTime() + ((long) task.getStartDelayInSeconds() * 1000L));
+        String reason = null;
```
Comment (andrea11) [inline core/src/main/java/com/netflix/conductor/core/execution/DeciderService.java]: maybe use Optional?

Comment (nbraquart) [reply core/src/main/java/com/netflix/conductor/core/execution/DeciderService.java]: No longer applicable

File: common/src/main/java/com/netflix/conductor/common/metadata/tasks/TaskDef.java (hunk 2 above)
Comment (nbraquart) [reply common/src/main/java/com/netflix/conductor/common/metadata/tasks/TaskDef.java]: Done

File: core/src/main/java/com/netflix/conductor/core/execution/DeciderService.java (hunk 1 above)
Comment (nbraquart) [reply core/src/main/java/com/netflix/conductor/core/execution/DeciderService.java]: Done

File: core/src/main/java/com/netflix/conductor/core/execution/DeciderService.java [hunk 4]
```diff
@@ -536,8 +537,36 @@ void checkTaskTimeout(TaskDef taskDef, Task task) {
...
+            LOGGER.warn("Missing task definition for task:{}/{} in workflow:{}", task.getTaskId(), task.getTaskDefName(), task.getWorkflowInstanceId());
+            return;
+        }
+        if (taskDef.getPollTimeoutSeconds() == null || taskDef.getPollTimeoutSeconds() <= 0 || !task.getStatus().equals(SCHEDULED)) {
+            return;
+        }
+
+        final long pollTimeout = 1000L * taskDef.getPollTimeoutSeconds();
+        final long now = System.currentTimeMillis();
+        final long pollElapsedTime = now - (task.getScheduledTime() + ((long) task.getStartDelayInSeconds() * 1000L));
```
Comment (apanicker-nflx) [inline core/src/main/java/com/netflix/conductor/core/execution/DeciderService.java]: When a task is sent back from the worker with `callbackAfterSeconds` set and `IN_PROGRESS` status, Conductor will mark this task as SCHEDULED and set it to invisible state in the queue to be available for polling after the specified callbackAfterSeconds. [Reference](https://github.com/Netflix/conductor/blob/master/core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java#L816-L819)

In these cases, the `pollElapsedTime` calculation here would wrongly mark this task to be poll timed out. This calculation would need to take into account, this scenario of callbackAfterSeconds and invisibility period on the queues. A similar time calculation can be found in the `isResponseTimedOut` method in this class for reference.

Comment (nbraquart) [reply core/src/main/java/com/netflix/conductor/core/execution/DeciderService.java]: Thanks for the feedback. I have added the callbackAfterSeconds to the timeout for a correct computation as per your suggestion.

**Pull request merged**